design/fast_finality.md

```diff
@@ -0,0 +1,276 @@
+  `Possible designs for preventing value-extraction`, below)
+* Do the light batches need to be linked to the latest block that was fed into the enclave?
+* How do we prevent denial-of-service attacks on the sequencer?
+
+## Appendices
+
+### Possible designs for preventing value-extraction
+
+In this section, we investigate various designs to prevent value-extraction. The specific attack we have in mind is one 
+where the sequencer runs `n` enclaves, and uses `n-1` of them to test the impact of various transaction sets to 
+identify value-extraction opportunities.
+
+#### Do nothing
+
+In this approach, we rely on trust in the sequencer and the fact that value-extraction opportunities are reduced in
+Obscuro due to its data-visibility rules.
+
+#### Disable `eth_call` on sequencer enclaves
+
+By disabling `eth_call` on the sequencer enclaves, we prevent the operator from extracting information about the impact
+of a given light batch.
+
+This approach is unworkable. The operator can run a separate, non-sequencer enclave, feed the light batches to that
+enclave, and use `eth_call`s on that enclave to determine the impact of a given light batch.
+
+#### Run a single sequencer enclave
+
+If there is a single sequencer enclave across the entire network, there are no other enclaves to use to identify 
+value-extraction opportunities. The single sequencer cannot be restarted to identify value-extraction opportunities, 
+since the enclave start-up delay will then prevent the sequencer from reaching its block production target.
+
+This approach is unworkable. We cannot achieve the desired recovery times with a single sequencer enclave.
+
+#### Detect restarts on sequencer enclaves
+
+We allow validators to detect how often sequencer enclaves are restarted, incentivising good behaviour on behalf of the 
+sequencer, and allowing the issue of a malicious (or incompetent) sequencer to be detected and handled as a governance 
+action. For this to work, validators would have to actively assess whether the sequencer is doing an adequate job.
+
+There are two flavours of this.
+
+##### Have sequencer enclaves produce lifetime proofs
+
+Every `x`th light batch includes a proof of how long (e.g. in terms of light batches or L1 blocks) each sequencer 
+enclave has been up. This creates a history of when each sequencer enclave was restarted. `x` can be arbitrarily high, 
+since you can work backwards from this proof and the previous proof to check the enclave has been up the entire time.
+
+An alternative model is for each sequencer enclave to post a restart proof to the L1 or light-batch chain whenever it 
+starts, and wait for that proof to be included in the chain before continuing execution (thus forcing the proof to be 
+posted on each restart).
+
+This history can be queried via RPC from validators.
+
+In this model, the sequencer operator would get `n-1` shots at front-running before having to restart one or more 
```

Comment:
I'd reinforce after this sentence something like: "And this abuse would be public and obvious to any observers."

Reply:
I don't think so - my understanding is that the `n-1` shots would be secret (provided the sequencer doesn't restart the nodes). But at the expense of the enclave being useless. I've clarified.

Reply:
My understanding (and this could be wrong) is that the validators would still be able to see that the LB has come from a different enclave. Each of the enclaves will have had to go through the secret exchange process which I guess will still be public and they'd each have their own key-pair to identify them.

Reply:
The `n-1` light batches would never be shared with anyone outside the sequencer, so no-one could see them and challenge them.

Reply:
Then the node operator can get no benefit, they've had to publish the LB they came up with first

Reply:
If you wanted to make it more robust you could force it to publish an empty "E's dead Jim, I'm taking over!" empty LB batch before they start processing transactions.

Reply:
"Then the node operator can get no benefit, they've had to publish the LB they came up with first" - I don't think so. They'd just continue publishing the LBs from the leader.

"If you wanted to make it more robust you could force it to publish an empty "E's dead Jim, I'm taking over!" empty LB batch before they start processing transactions." - but the sequencer operator could just intercept this before it's broadcast to the broader network.

Reply:
When the leader enclave returns an LB from a submit block it has to publish that or die. It can't publish a different LB at the same height. So if you want plausible deniability you could just give yourself 1 more roll of the dice and pretend it was a real failover, if you don't like the second roll of the dice you're publishing multiple failovers to the chain.